src/bin/languageserver/mod.rs

```diff
@@ -1793,45 +1899,158 @@ impl LanguageServer for SolangServer {
+            return Ok(None);
+        };
+        let definitions = &self.global_cache.lock().await.definitions;
+        let location = definitions
+            .get(&reference)
+            .map(|range| {
+                let uri = Url::from_file_path(&reference.def_path).unwrap();
+                Location { uri, range: *range }
+            })
+            .map(GotoTypeDefinitionResponse::Scalar);
+        Ok(location)
+    }
 
-        let path = uri.to_file_path().map_err(|_| Error {
-            code: ErrorCode::InvalidRequest,
-            message: format!("Received invalid URI: {uri}").into(),
-            data: None,
-        })?;
-        let files = self.files.lock().await;
-        if let Some(cache) = files.caches.get(&path) {
-            let f = &cache.file;
-            let offset = f.get_offset(
-                params.text_document_position_params.position.line as _,
-                params.text_document_position_params.position.character as _,
-            );
-            if let Some(reference) = cache
-                .references
-                .find(offset, offset + 1)
-                .min_by(|a, b| (a.stop - a.start).cmp(&(b.stop - b.start)))
-            {
-                let di = &reference.val;
-                if let Some(range) = self.definitions.lock().await.get(di) {
-                    let uri = Url::from_file_path(&di.def_path).unwrap();
-                    let ret = Ok(Some(GotoDefinitionResponse::Scalar(Location {
-                        uri,
-                        range: *range,
-                    })));
-                    return ret;
+    async fn goto_type_definition(
+        &self,
+        params: GotoTypeDefinitionParams,
+    ) -> Result<Option<GotoTypeDefinitionResponse>> {
+        let Some(reference) = self.get_reference_from_params(params).await? else {
+            return Ok(None);
+        };
+        let gc = self.global_cache.lock().await;
+
+        let di = match &reference.def_type {
+            DefinitionType::Variable(_)
+            | DefinitionType::NonLocalVariable(_, _)
+            | DefinitionType::Field(_, _)
+            | DefinitionType::Variant(_, _) => {
+                if let Some(def) = gc.types.get(&reference) {
+                    def
+                } else {
+                    return Ok(None);
                 }
             }
+            DefinitionType::Struct(_)
+            | DefinitionType::Enum(_)
+            | DefinitionType::Contract(_)
+            | DefinitionType::Event(_)
+            | DefinitionType::UserType(_) => &reference,
+            _ => return Ok(None),
+        };
+
+        let location = gc
+            .definitions
+            .get(di)
+            .map(|range| {
+                let uri = Url::from_file_path(&di.def_path).unwrap();
+                Location { uri, range: *range }
+            })
+            .map(GotoTypeDefinitionResponse::Scalar);
+
+        Ok(location)
+    }
+
+    async fn goto_implementation(
```

Comment:
Needs a doc comment explaining the return type and what the function does on a high level